### PR TITLE
Replace most of the cacheData with RCBytes

### PR DESCRIPTION
**Pull request by nnsgmsone**
## What type of PR is this?

- [ ] API-change
- [X] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #10945

## What this PR does / why we need it:
* This pr is the first pr to fix the memory problem, which replaces most of the cachedata with rcbytes
* Change rcbyte from globally shared to one per fileservice

## this PR undo
* metadata still doesn't use rcbytes at the moment, metadata's memory will currently be held by two lru's, the lifecycle is a bit messed up, no fixes at the moment
* Memory allocator and lru are currently still cut off and can't limit the memory size very well

### CODE REVIEW

File: pkg/fileservice/rc_bytes.go [hunk 1]
```diff
@@ -14,92 +14,92 @@
...
+func alloc(size int) []byte {
+	ptr := C.calloc(C.size_t(size), 1)
+	if ptr == nil {
+		// NB: throw is like panic, except it guarantees the process will be
+		// terminated. The call below is exactly what the Go runtime invokes when
+		// it cannot allocate memory.
+		panic("out of memory")
+	}
+	// Interpret the C pointer as a pointer to a Go array, then slice.
+	return (*[MaxArrayLen]byte)(unsafe.Pointer(ptr))[:size:size]
```
Comment (reusee) [inline pkg/fileservice/rc_bytes.go:95]: should use unsafe.Slice((*byte)(unsafe.Pointer(ptr)), size)

File: pkg/fileservice/rc_bytes.go [hunk 2]
```diff
@@ -14,92 +14,92 @@
...
-
-func (r *rcBytesPool) Alloc(size int) CacheData {
-	return r.Get(size)
+// free frees the specified slice.
+func free(b []byte) {
+	if cap(b) != 0 {
+		b = b[:cap(b)]
+		ptr := unsafe.Pointer(&b[0])
+		C.free(ptr)
+	}
```
Comment (reusee) [inline pkg/fileservice/rc_bytes.go:104]: should be if ptr := unsafe.SliceData(b); ptr != nil { C.free(unsafe.Pointer(ptr)) }

File: pkg/objectio/funcs.go [hunk 3]
```diff
@@ -37,13 +37,13 @@ func ReadExtent(
 	ioVec := &fileservice.IOVector{
 		FilePath:    name,
 		Entries:     make([]fileservice.IOEntry, 1),
-		CachePolicy: cachePolicy,
+		CachePolicy: fileservice.SkipMemory,
```
Comment (reusee) [inline pkg/objectio/funcs.go:40]: why skip memory here?

**Review by reusee — CHANGES_REQUESTED** [review]

File: pkg/fileservice/rc_bytes.go (hunk 1 above)
Comment (nnsgmsone) [reply pkg/fileservice/rc_bytes.go:95]: 这个应该就是对的用法。。岳声

File: pkg/objectio/funcs.go (hunk 3 above)
Comment (nnsgmsone) [reply pkg/objectio/funcs.go:40]: metadata现在缓存了两份阿。。为啥需要两份呢。。。完全没必要的。。两个lru，所以这里跳过了

File: pkg/fileservice/rc_bytes.go (hunk 2 above)
Comment (nnsgmsone) [reply pkg/fileservice/rc_bytes.go:104]: 这个应该就是对的用法。。岳声

File: pkg/objectio/funcs.go (hunk 3 above)
Comment (LeftHandCold) [reply pkg/objectio/funcs.go:40]: Here the "skipmemory" parameter should be passed in from the outside

File: pkg/fileservice/mem_cache_test.go [hunk 4]
```diff
@@ -64,7 +64,7 @@ func TestMemCacheLeak(t *testing.T) {
 	assert.Nil(t, err)
 	err = m.Update(ctx, vec, false)
 	assert.Nil(t, err)
-	assert.Equal(t, int64(1), m.cache.Capacity()-m.cache.Available())
+	assert.Equal(t, int64(0), m.cache.Capacity()-m.cache.Available())
```
[the remark this answers is not in the document]
Comment (nnsgmsone) [reply pkg/fileservice/mem_cache_test.go:67]: 实现的了，唯一的差别是size是使用的rcbytes的size，这样才方便控制内存

File: pkg/fileservice/io_entry.go [hunk 5]
```diff
@@ -27,21 +27,23 @@ func (e *IOEntry) setCachedData() error {
...
+func (e *IOEntry) ReadFromOSFile(file *os.File, alloc CacheDataAllocator) error {
 	r := io.LimitReader(file, e.Size)
 
-	if len(e.Data) < int(e.Size) {
-		e.Data = make([]byte, e.Size)
-	}
-
+	cacheData := alloc.Alloc(int(e.Size))
+	defer func() {
+		e.Data = nil
```
Comment (reusee) [inline pkg/fileservice/io_entry.go:43]: the caller needs the Data field, so should not be set to nil.

**Review by reusee — CHANGES_REQUESTED** [review]

**Pull request closed without merging**